// hd-wallet/lib/build-tx/result.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getNonfinalResult = getNonfinalResult;
exports.getFinalResult = getFinalResult;
exports.empty = void 0;

var mtransaction = _interopRequireWildcard(require("./transaction"));

var coinselect = _interopRequireWildcard(require("./coinselect"));

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj["default"] = obj; return newObj; } }

var empty = {
  type: 'error',
  error: 'EMPTY'
};
exports.empty = empty;

function getNonfinalResult(result) {
  var _result$result = result.result,
      max = _result$result.max,
      fee = _result$result.fee,
      feePerByte = _result$result.feePerByte,
      bytes = _result$result.bytes,
      totalSpent = _result$result.totalSpent;
  return {
    type: 'nonfinal',
    fee: fee,
    feePerByte: feePerByte,
    bytes: bytes,
    max: max,
    totalSpent: totalSpent
  };
}

function getFinalResult(result, transaction) {
  var _result$result2 = result.result,
      max = _result$result2.max,
      fee = _result$result2.fee,
      feePerByte = _result$result2.feePerByte,
      bytes = _result$result2.bytes,
      totalSpent = _result$result2.totalSpent;
  return {
    type: 'final',
    fee: fee,
    feePerByte: feePerByte,
    bytes: bytes,
    transaction: transaction,
    max: max,
    totalSpent: totalSpent
  };
}
